exit/test_exit.c

```diff
@@ -0,0 +1,518 @@
+/*
+ * Phoenix-RTOS
+ *
+ *    POSIX.1-2017 standard library functions tests
+ *    HEADER:
+ *    - stdlib.h
+ *    - unistd.h
+ *    TESTED:
+ *    - exit()
+ *    - _exit() (_Exit equivalent)
+ *    - atexit()
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Adam Debek
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unity_fixture.h>
+#include "common.h"
+
+#define PATH       "/tmp/testFile"
+#define CREATEMODE 0666
+#define TEST_STR   "test123"
+
+static int fd;
+static int val = 0;
+
+
+/* SIGCHLD signal handler */
+static void handler(int signum)
+{
+	if (signum != SIGCHLD) {
+		TEST_MESSAGE("Handler received other signal than SIGCHLD");
+		_exit(EXIT_FAILURE);
+	}
+
+	if (creat(PATH, CREATEMODE) == -1) {
+		TEST_MESSAGE("Couldn't create file in handler");
+		_exit(EXIT_FAILURE);
+	}
+}
+
+
+/* Functions to be registered by atexit */
+static void fun1(void)
+{
+	val++;
+	TEST_ASSERT_EQUAL_INT(sizeof(int), write(fd, &val, sizeof(int)));
+}
+
+
+static void fun2(void)
+{
+	int val = 40;
+	TEST_ASSERT_EQUAL_INT(sizeof(int), write(fd, &val, sizeof(int)));
+}
+
+
+TEST_GROUP(_exit);
+
+
+TEST_SETUP(_exit)
+{
+}
+
+
+TEST_TEAR_DOWN(_exit)
+{
+}
+
+
+TEST(_exit, status_vals)
+{
+	/* Return all possible least significant byte values */
+	pid_t pid;
+	int status, i;
+	int val[3];
+	val[0] = 0x1 << 8;
+	val[1] = (0x1 << 16) + 1;
+	val[2] = (0x1 << 24) + 2;
```

Comment:
This is done because only the least significant 8 bits shall be available to a waiting parent process. 3 most significant bytes should be cut off so returning val[0] should resolve to returning 0, val[1] to 1 and val[2] to 2.